### ISSUE TITLE
Obtain per-class Metrics relevant to training if the cell output is lost

### ISSUE THREAD

**Issue by emailic** [ultralytics/yolov5#8667]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, 

I would like to know how to obtain the table with metrics (precision, recall, mAP..) per class. 
I know it is printed at the end of training, but what happens if I lose cell output? Is there a way to re-obtain this information? 

On that note,  recently I resumed training starting epoch 95, using the last weights saved and at the end of training
 (not using the 'resume' flag because I was having problems running the code that way), and at the end of the training, the  table with metrics per class was not printed. 
How can I obtain it?
Just to be clear, I would like to obtain the metrics relevant to the training and **validation**, not to the test set. 

`torch.cuda.empty_cache()`

`!python '/content/drive/My Drive/microplasticos/yolov5-master/train.py'  --batch 14 --epochs 250 --data '/content/drive/My Drive/microplasticos/microplasticos_576.yaml' --patience 40 \
--weights  '/content/drive/My Drive/microplasticos/microplasticos_576/feature_extraction14/weights/last.pt' --project '/content/drive/My Drive/microplasticos/microplasticos_576'  --imgsz=576 \
--name 'feature_extraction' --freeze 12 --cache ram`

![image](https://user-images.githubusercontent.com/83206159/180279044-99f0552a-968f-47bb-9407-de1555a26213.png)

[In this issue, ](https://github.com/ultralytics/yolov5/issues/5880) i saw @kristofferK trying to obtain something similar, though I didn't fully understand what was he trying to obtain. So I tested his solution on another jupyter notebook where I actually still have the data printed in the cell output. The new metrics results didn't match the previous ones, as you can see in the screenshot below. Could you give me a hand in understanding why are the results different? 
![image](https://user-images.githubusercontent.com/83206159/180282347-f099f37a-4917-46d1-bb69-10b9887b7804.png)

**Comment by glenn-jocher** [comment]
@emailic 👋 Hello! Thanks for asking about **resuming training**. val.py will print per class metrics, i.e.
```
python val.py --data ... --weights ...
```

YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/aa542ce6a65658ff931fee9bbab77c0145c152f0/train.py#L175-L181) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="Screenshot 2022-04-10 at 11 11 51" src="https://user-images.githubusercontent.com/26833433/162611202-4e0b611f-6cca-4e43-94fc-bbc44e6447f8.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck 🍀 and let us know if you have any other questions!

**Comment by emailic** [comment]
Hi @glenn-jocher , thanks for getting back to me!

To address your answer from the beginning:
>'python val.py --data ... --weights ... will  print per class metrics.'

Indeed, it does print per class metrics, however, these per class metrics are not the same as the metrics printed at the end of training, as can be seen in the last (second) screenshot of the original issue. Thus I'm curious to know why is it so? 

Secondly,
as you advised, I tried resuming by executing 'python train.py --resume', and then  'python train.py --resume path/to/last.pt'. However, as you can see on the screenshot below, that initiated a coco dataset and yolov5s weights, low augmentation settings, image size 640, etc.... all the settings that were not a part of the last training. The real settings of the last training (which was interrupted) are the following: 

`!python train.py  --batch 8 --epochs 150 --hyp hyp.scratch-high.yaml --data '/content/drive/My Drive/microplasticos/microplasticos_1152.yaml' --weights yolov5s6.pt --project '/content/drive/My Drive/microplasticos/microplasticos_1152'  --imgsz=1152 --name 'feature_extraction'  --cache ram
`

![image](https://user-images.githubusercontent.com/83206159/180757211-2a507d66-8bc4-4b50-9ed6-0550f4588f18.png)

This made me specify the other flags:

`!python train.py  --batch 8 --epochs 150 --hyp hyp.scratch-high.yaml --data '/content/drive/My Drive/microplasticos/microplasticos_1152.yaml' --weights yolov5s6.pt --project '/content/drive/My Drive/microplasticos/microplasticos_1152'  --imgsz=1152 --name 'feature_extraction'  --cache ram --resume True
`
And have gotten an error message.
![image](https://user-images.githubusercontent.com/83206159/180758287-39aba0ac-0f62-45b0-9f0b-6d5a51fb4362.png)

I corrected the 'feature_extraction' flag to feature_extraction21, which is the last feature extraction (the interrupted one), and I received the same error message again. 
Thus I'm still not managing to resume the last training. Any advice?

Thanks for your patience @glenn-jocher 
Take care, Ema

**Comment by glenn-jocher** [comment]
@emailic --resume operates correctly. I just tested this in Google Colab, resume starts from the correct checkpoint and logs results to the right directory. I'm not able to reproduce any issues.

```python
!git clone https://github.com/ultralytics/yolov5  # clone
%cd yolov5
%pip install -qr requirements.txt  # install

import torch
import utils
display = utils.notebook_init()  # checks

# Reproduce
!python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --cache 
!python train.py --img 320 --batch 32 --epochs 10 --data coco128.yaml --weights yolov5m.pt --cache  # CTRL+C
!python train.py --resume
```

<img width="1307" alt="Screen Shot 2022-07-29 at 5 42 37 PM" src="https://user-images.githubusercontent.com/26833433/181795499-8187eb22-7c60-4feb-a914-c363e0fd9284.png">

**Comment by emailic** [comment]
Hi @glenn-jocher !
Thanks for getting back to me.
Have you tried restarting the  kernel before executing !python train.py --resume? 
Take care, Ema

**Comment by glenn-jocher** [comment]
@emailic 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃
